File: 05_18_week12/EXERCISE/CPP/stack.cpp
#include <iostream>

#define MAX_STACK_SIZE 3

typedef int element;
typedef struct{
    element stack[MAX_STACK_SIZE];
    int top;
}Stack;

void init(Stack *stack){
    stack->top = -1;
}

bool is_empty(Stack *stack){
    return stack->top == -1;
}

bool is_full(Stack *stack){
    return stack->top == MAX_STACK_SIZE-1;
}

void push(Stack *stack, element data){
    if(is_full(stack)){
        std::cerr << "Stack is FULL\n";
        exit(-1);
    }
    stack->stack[++stack->top] = data;
}

element pop(Stack *stack){
    if (is_empty(stack)){
        std::cerr << "Stack is Empty\n";
        exit(-1);
    }
    return stack->stack[stack->top--];
}

element peek(Stack *stack){
    if (is_empty(stack)){
        std::cerr << "Stack is Empty\n";
        exit(-1);
    }
    return stack->stack[stack->top];
}

int main(){
    Stack stack;    // 스택 객체 선언

    init(&stack);    // 초기화
    push(&stack, 1);
    push(&stack, 2);
    push(&stack, 3);    // 스택 삽입

    std::cout << is_full(&stack) << "\n";
    std::cout << pop(&stack) << "\n";
    std::cout << pop(&stack) << "\n";
    std::cout << peek(&stack) << "\n";
    std::cout << pop(&stack) << "\n";
    std::cout << is_empty(&stack) << "\n";
}

File: 05_18_week12/EXERCISE/Python/Stack.py
MAX_STACK_SIZE = 3                  # stack의 최대 크기

class Stack:
    def __init__(self):             # stack 초기화
        self.stack = [0] * MAX_STACK_SIZE
        self.top = -1
        
    def is_empty(self):             # stack이 비어있는지 확인
        return self.top == -1
    
    def is_full(self):              # stack이 꽉 차 있는지 확인
        return self.top == MAX_STACK_SIZE - 1
    
    def push(self, data):           # stack에 data 삽입
        if self.is_full():
            print("Stack is Full")
            exit(-1)
        self.top += 1
        self.stack[self.top] = data
        
    def pop(self):                  # stack에서 data 반환하고 삭제
        if self.is_empty():
            print("Stack is Empty")
            exit(-1)
        return self.stack[self.top]
    
    def peek(self):                 # stack에서 data 반환하고 삭제 X
        if self.is_empty():
            print("Stack is Empty")
            exit(-1)
        return self.stack[self.top]
    
if __name__ == "__main__":
    stack = Stack()
    
    stack.push(1)
    stack.push(2)
    stack.push(3)

    print(stack.is_full())      # True
    print(stack.pop())          # 3
    print(stack.pop())          # 2
    print(stack.peek())         # 1
    print(stack.pop())          # 1
    print(stack.is_empty())     # True